Sb rerefactor

danamansana opened the pull request:
This branch adds caching.

danamansana commented:
We should look into handling the potential situation in which Redis fails while the app is running. Currently this will break everything. In the old PR, da_sb_caching, Kang asked what would happen in checkKeyForRedis if the promise did not resolve. I believe the only way this could happen is if Redis fails, which currently will cause a much larger problem

thisisstephenbetts commented:
Ooh, good thought. Yeah, we should catch that error and go to synchronous calls.

ktp242 commented on README.md:
```diff
@@ -33,6 +33,13 @@ You do not need any environment variables if you want to run it locally in devel
 `npm start`
 
+You will need to have a redis server running in another tab if you want to use caching:
+`redis-server`
```
I am not sure I understand the command here. Should I do `npm run redis-server` in another tab? The in the same path?

danamansana replied:
You should just do `redis-server` in another tab

ktp242 commented on test/cacheUtil.test.js:
```diff
@@ -0,0 +1,163 @@
+import cacheMethods from '../src/server/ApiRoutes/cacheUtil';
+
+const chai = require('chai');
```
You can use ES6 way to import modules as aligning with other files.
`import chai from 'chai';`

danamansana replied:
Fixed!

ktp242 commented on test/cacheUtil.test.js:
```diff
@@ -0,0 +1,163 @@
+  describe('checkForKeyInRedis', () => {
+    it('should resolve to a value that has been set', () => {
+      mockClient.set('hello', 'world');
+      return expect(checkForKeyInRedis('hello'))
```
I am not sure why we need to return the function in a test.

danamansana replied:
I think it has to do with the way chai deals with asynchronous behavior, it will not register a failure before the it block completes unless expect is returned

ktp242 commented:
I think we accidentally merge `server/ApiRoutes/cacheUtil.js` in to `pr_approved` lol We should be more careful next time.

Also, it is minor, but the other file names in `server/ApiRoutes` are capitalized, it would be better to keep them aligned. So are the file names in `/test`

ktp242 commented on src/server/ApiRoutes/cacheUtil.js:
```diff
@@ -1,27 +1,104 @@
+    return Promise.resolve(dataFunction);
+  }
+
+  return (process.env.APP_ENV
```
Can we pass `process.env.APP_ENV` as a parameter to `addCaching`? Like you did with `useClient `, which is the value of `process.env.SKIP_CACHING`?

danamansana replied:
OK

ktp242 commented on src/server/ApiRoutes/cacheUtil.js:
```diff
@@ -1,27 +1,104 @@
+  return (process.env.APP_ENV
+    ? kms.decrypt(redisHosts[process.env.APP_ENV], null, 'us-east-1')
+      .then(data => new ClientWrapper(data))
+    : Promise.resolve(new ClientWrapper()))
```
I am not very sure about the logic here.

So what I see is, if we have `process.env.APP_ENV`, we utilize `kms.decrypt` to get the Redis Host endpoint and then construct a new client.

If we don't, we still make a new client, but then we check if a client is already there.

Shouldn't we first check if a client is already passed as a parameter and run `useCachedOrGetData(dataFunction, client)` to get data?

And then if no client passed, we create a new one with  `new ClientWrapper()`

danamansana replied:
I've factored this out into a separate method, hopefully it makes more sense now. In the new method, if passed a customClient, it does nothing, otherwise it checks the appEnv to see whether the clientWrapper needs a redis address, and returns a clientWrapper pointing to the address if there is one

danamansana commented on src/server/ApiRoutes/cacheUtil.js:
```diff
@@ -1,27 +1,104 @@
+  }
+
+  return (process.env.APP_ENV
+    ? kms.decrypt(redisHosts[process.env.APP_ENV], null, 'us-east-1')
```
OK

ktp242 commented on src/server/ApiRoutes/ApiRoutes.js:
```diff
@@ -1,14 +1,20 @@
-
-const getSearchData = url => axios.get(url);
+const { addCaching } = cache;
+let getSearchData;
```
I think line 13 and 14 can be combined.

danamansana replied:
Not sure what you mean here

ktp242 replied:
I was thinking we can make `addCaching` a pure function, that is it doesn't mutate a variable but simply returns a value. It will be an improvement but I think it is OK for now.

ktp242 commented on src/server/ApiRoutes/cacheUtil.js:
```diff
@@ -1,27 +1,117 @@
+  }
+
+  return generateClient(customClient, appEnv, region)
+    .then(client => (...params) => useCachedOrGetData(dataFunction, client)(params)
```
Where do we indicate all the parameters here? I guess I am a bit confused with the workflow here

danamansana replied:
Do you want to discuss this?

danamansana commented:
Following the suggestions above, I've: added comma delimiters to the redis key; updated the comments for cacheUtil functions to match the new params;  and fixed the import statement for redisHost. I'll add something about redis to the README.

danamansana commented on src/server/ApiRoutes/cacheUtil.js:
```diff
@@ -1,27 +1,117 @@
-  const checkForKeyInClient = key => new Promise((resolve) => {
-    client.get(key, (err, cachedValue) => resolve(cachedValue));
-  });
+const getKeyFromParams = params => params.map(x => JSON.stringify(x)).join('');
```
Good idea, fixed!

danamansana commented on src/server/ApiRoutes/cacheUtil.js:
```diff
@@ -1,27 +1,117 @@
-import redis from 'redis';
+import appConfig from '../../../appConfig';
```
Fixed

danamansana commented on src/server/ApiRoutes/cacheUtil.js:
```diff
@@ -1,27 +1,117 @@
+
+ @param {function} datafunction
+ @param {boolean} useClient
+ @param {object} customClient
```
Fixed

danamansana commented on test/cacheUtil.test.js:
```diff
@@ -0,0 +1,158 @@
+
+    values = {};
+
+    mockClient = {
```
OK, I've refactored this to be a bit more OO and got rid of the values variable

danamansana commented on .travis.yml:
```diff
@@ -50,7 +50,7 @@ deploy:
   on:
     repo: NYPL/dgx-global-search
-    branch: development
+    branch: sb_rerefactor
```
Fixed

danamansana commented on src/server/ApiRoutes/clientWrapper.js:
```diff
@@ -0,0 +1,37 @@
+import redis from 'redis';
+
+/**
+ ClientWrapper:
```
Added!

danamansana commented on src/server/ApiRoutes/cacheUtil.js:
```diff
@@ -1,27 +1,104 @@
+ to see if those params are cached, and then either returns the cached value or gets
+ a new value for those params if none is cached
+
+ @param {function} dataFunction
```
Sorry, not sure what you want to do here

ktp242 replied:
oooops sorry I think I saw something out of date.

But if you check line 90, you will see it has 2 parameters while on line 104, it has 5.

danamansana commented on src/server/ApiRoutes/cacheUtil.js:
```diff
@@ -1,27 +1,116 @@
+ @param {string} key
+ */
+
+const getDataAndSetKeyInClient = (dataFunction, client) => (params, key) => dataFunction(...params)
```
Yeah this is mostly just to make testing easier

danamansana commented:
I've updated the CHANGELOG
